### PR TITLE
INSIGHT-1230: implementing deduplication endpoints

### PR DESCRIPTION
**Pull request by andreinechaev**

### CODE REVIEW

File: insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    "Failed to index document {} with xpath {} for project {};\nURL: {}\nresponse code: {}\nmessage: {}",
+                    docId, xpath, client.getProjectId(), client.getUrl(), response.code(), response.message());
+            return false;
+        });
+    }
+
+    private List<String> handleFind(Map<String, Serializable> parameters, TensorInstances payload)
+            throws JsonProcessingException {
+        String docId = (String) parameters.get(UID);
+        String xpath = (String) parameters.getOrDefault(XPATH_PARAM, DEFAULT_XPATH);
```
Comment (PedroCardoso) [inline insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java:109]: I really would prefer to always force a XPATH at this level. Define a default on ai-core if needed, but this stay open.

File: insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            return (T) handleFind(parameters, (TensorInstances) payload);
+        default:
+            throw new InvalidEndpointException("No such endpoint " + this.type.name());
+        }
+    }
+
+    private Boolean handleIndex(Map<String, Serializable> parameters, TensorInstances payload)
+            throws JsonProcessingException {
+        String docId = (String) parameters.get(UID);
+        String xpath = (String) parameters.getOrDefault(XPATH_PARAM, DEFAULT_XPATH);
```
Comment (PedroCardoso) [inline insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java:86]: I really would prefer to always force a XPATH at this level. Define a default on ai-core if needed, but this stay open.

File: insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @SuppressWarnings("unchecked")
+    public <T> T call(Map<String, Serializable> parameters, Serializable payload) throws IOException {
+        switch (this.type) {
+        case INDEX:
+            return (T) handleIndex(parameters, (TensorInstances) payload);
+        case FIND:
+            return (T) handleFind(parameters, (TensorInstances) payload);
+        default:
+            throw new InvalidEndpointException("No such endpoint " + this.type.name());
+        }
```
Comment (PedroCardoso) [inline insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java:80]: How about iterate tuples of similar documents ?

File: pom.xml [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
   <groupId>org.nuxeo.ai</groupId>
   <artifactId>nuxeo-insight-client</artifactId>
-  <version>1.1.1-SNAPSHOT</version>
+  <version>1.2.0-SNAPSHOT</version>
```
Comment (PedroCardoso) [inline pom.xml:7]: just a note for the future, this should be a PR on its own.

File: insight-rest-client/src/test/java/org/nuxeo/ai/sdk/rest/api/TestDedupCaller.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Test
+    public void shouldFindBasedOnTensor() throws IOException {
+        InsightClient client = getInsightClient();
+        HashMap<String, Serializable> params = new HashMap<>();
+        params.put(UID, "document_uuid_001");
+        params.put(XPATH_PARAM, DEFAULT_XPATH);
+
+        List<String> result = client.api(Dedup.FIND).call(params, createTensor("document_uuid_001"));
```
Comment (PedroCardoso) [inline insight-rest-client/src/test/java/org/nuxeo/ai/sdk/rest/api/TestDedupCaller.java:81]: If we are searching for similar based on a tensor (image) we should not need a UID and XPATH as they might not even exist.

**Review by PedroCardoso — CHANGES_REQUESTED** [review]
Requesting changes, due to the test with a tensor, without UID and xpath parameters.

File: insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java (hunk 1 above)
Comment (andreinechaev) [reply insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java:109]: ok

File: pom.xml (hunk 4 above)
Comment (andreinechaev) [reply pom.xml:7]: ok

**Pull request merged**